node/src/components/consensus/era_supervisor.rs

```diff
@@ -228,12 +230,45 @@ where
         result.into()
     }
 
+    fn booking_block_height(&self, era_id: EraId) -> u64 {
+        // The booking block for era N is the last block of era N - AUCTION_DELAY - 1
+        // To find it, we get the start height of era N - AUCTION_DELAY and subtract 1
+        let after_booking_era_id = EraId(era_id.0.saturating_sub(AUCTION_DELAY));
+        self.active_eras
+            .get(&after_booking_era_id)
+            .expect("should have era after booking block")
+            .start_height
+            .saturating_sub(1)
+    }
+
+    fn key_block_height(&self, _era_id: EraId, start_height: u64) -> u64 {
+        // the switch block of the previous era
+        // TODO: consider defining the key block as a block further in the past
+        start_height.saturating_sub(1)
```

Comment:
Era ID should be enough, right? You can calculate the height of the key block given an era in which it was created.

Reply:
It should... but if the era ID points to the era we are currently creating, we don't yet have it stored, and so we can't read its starting height from anywhere. That's why for this specific case, I need to pass the starting height of the new era as well.

This might be made obsolete if we decide to make the key block be further in the past.

Reply:
Wouldn't `era_id.saturating_sub(1).start_height` work?

Reply:
But the `start_height` of the previous era is the first block of that era, not the last. And there is no easy way to find out how many blocks were there in that era, so we can't even add era length to its `start_height`...

Reply:
Aha, thanks for the explanation. Makes sense.